orchagent/copporch.cpp

```diff
@@ -0,0 +1,488 @@
+        if (!getTrapID(trap_id_str, trap_id))
+        {
+            SWSS_LOG_ERROR("Invalid trap_id value specified %s\n", trap_id_str.c_str());
+            return false;
+        }
+        SWSS_LOG_DEBUG("Input trap_id:%s, pushing trap_id:%d", trap_id_str.c_str(), trap_id);
+        trap_id_list.push_back(trap_id);
+    }
+    return true;
+}
+
+bool CoppOrch::applyTrapIds(sai_object_id_t trap_group, vector<string> &trap_id_name_list, std::vector<sai_attribute_t> &trap_id_attribs)
+{
+    SWSS_LOG_ENTER();
+    sai_status_t sai_status;
+    vector<sai_hostif_trap_id_t> trap_id_list;
+    if (!getTrapIdList(trap_id_name_list, trap_id_list))
+    {
+        return false;
+    }
+
+    sai_attribute_t attr;
+    attr.id = SAI_HOSTIF_TRAP_ATTR_TRAP_GROUP;
+    attr.value.oid = trap_group;
+    trap_id_attribs.push_back(attr);
+
+    // The channel is always the NETDEV device.
+    attr.id = SAI_HOSTIF_TRAP_ATTR_TRAP_CHANNEL;
+    attr.value.s32 = SAI_HOSTIF_TRAP_CHANNEL_NETDEV;
+    trap_id_attribs.push_back(attr);
+
+    for (auto trap_id : trap_id_list)
+    {
+        SWSS_LOG_DEBUG("set trap attr.1:%d", attr.id);
+        for(auto trap_id_attr : trap_id_attribs)
+        {
+            SWSS_LOG_DEBUG("set trap attr.2:%d", trap_id_attr.id);
+            if (SAI_STATUS_SUCCESS != (sai_status = sai_hostif_api->set_trap_attribute(trap_id, &trap_id_attr)))
+            {
+                SWSS_LOG_ERROR("Failed to apply trap_id attribute:%d to trap_id:%d, error:%d\n", trap_id_attr.id, trap_id, sai_status);
+                return false;
+            }
+        }
+    }
+    return true;    
+}
+
+bool CoppOrch::removePolicerFromTrapGroup(string trap_group_name)
+{
+    SWSS_LOG_ENTER();
+    sai_attribute_t attr;
+    sai_status_t sai_status;
+    sai_object_id_t policer_id = getTrapGroupPolicer(trap_group_name);
+    if (SAI_NULL_OBJECT_ID == policer_id)
+    {
+        SWSS_LOG_DEBUG("No policer is attached to trap group:%s\n", trap_group_name.c_str());
+        return true;
+    }
+    attr.id = SAI_HOSTIF_TRAP_GROUP_ATTR_POLICER;
+    attr.value.oid = SAI_NULL_OBJECT_ID;
+    if (SAI_STATUS_SUCCESS != (sai_status = sai_hostif_api->set_trap_group_attribute(m_trap_group_map[trap_group_name], &attr))) {
+        SWSS_LOG_ERROR("Failed to reset policer for trap group:%llx, error:%d\n", m_trap_group_map[trap_group_name], sai_status);
+        return false;
+    }
+    if (SAI_STATUS_SUCCESS != (sai_status = sai_policer_api->remove_policer(policer_id)))
+    {
+        SWSS_LOG_ERROR("Failed to remove policer:%llx, error:%d\n", policer_id, sai_status);
+        return false;
+    }
+    m_trap_group_policer_map.erase(m_trap_group_map[trap_group_name]);    
+    return true;
+}
+
+sai_object_id_t CoppOrch::getTrapGroupPolicer(string key)
+{
+    SWSS_LOG_ENTER();
+    SWSS_LOG_DEBUG("trap group name:%s:", key.c_str());
+    if (m_trap_group_map.find(key) == m_trap_group_map.end())
+    {
+        return SAI_NULL_OBJECT_ID;
```

Comment:
This is 'null' value, not error code. it's used by caller function to detect whether there is cached object.